src/SimCan.cpp

```diff
@@ -15,53 +15,106 @@ void SimCan::begin(){
+    if (millis() - _last_transmit >= CAN_TRANSMIT_INTERVAL) {
+        _last_transmit = millis();
+        _transmit();
     } 
 
-    // Print any CAN messages received (debug mode only)
+    // Listen for CAN messages
+    if (_can->checkReceive() == CAN_MSGAVAIL) {
+        _receive();
+    }
+
+}
+
+void SimCan::_transmit() {
+        for(CanMessage msg : TRANSMIT_MSGS){
+            uint8_t error = _can->sendMsgBuf(msg.id, CAN_FRAME, msg.dataLength, msg.data);
+            if(_serial){
+                _serial->print("CAN MESSAGE SENT - ID: 0x"); 
+                _serial->print(msg.id, HEX); 
+                _serial->println(" - Status: " + _getErrorDescription(error));
+            }
+        }
+
+}
+
+void SimCan::_receive(){
+    uint8_t len = 0;
+    uint8_t buf[8];
+
+    _can->readMsgBuf(&len, buf); 
+
+    uint16_t canId = _can->getCanId();
+
+    if(canId == BMS_REQUEST_ID){
+        _processBmsRequest(len, buf);
+    }
+
     if(_serial){
-        unsigned char len = 0;
-        unsigned char buf[8];
+        _serial->println("-----------------------------");
+        _serial->print("CAN MESSAGE RECEIVED - ID: 0x");
+        _serial->println(canId, HEX);
 
-        if (_can->checkReceive() == CAN_MSGAVAIL) {
-            _can->readMsgBuf(&len, buf); 
+        for (int i = 0; i < len; i++) { // print the data
+            _serial->print("0x");
+            _serial->print(buf[i], HEX);
+            _serial->print("\t");
+        }
+        _serial->println();
+    }
+}
 
-            unsigned long canId = _can->getCanId();
+void SimCan::_processBmsRequest(uint8_t len, uint8_t buf[]){
 
-            _serial->println("-----------------------------");
-            _serial->print("Get data from ID: 0x");
-            _serial->println(canId, HEX);
+    if(_serial) _serial->println("BMS CAN REQUEST RECEIVED!");
+    
+    // Check to make sure request length is correct
+    if (len != 8 && _serial){
+        _serial->println("ERROR: BMS CAN REQUEST TOO SHORT!");
+    }
 
-            for (int i = 0; i < len; i++) { // print the data
-                _serial->print(buf[i], HEX);
-                _serial->print("\t");
+    bool reqFulfilled = false;
+    // Check all the BMS properties that we're interested in and see if the first byte matches its id
+    for(uint8_t i = 0; i < NUM_BMS_PROPERTIES; i++){
+        if(buf[0] == BMS_PROPERTY[i]){
+            reqFulfilled = true;
+            // Create a response
+            uint8_t response[6];
+            // First byte is 0x01 to indicate OK
+            response[0] = 0x01;
+            // Second byte is the BMS property id
+            response[1] = BMS_PROPERTY[i];
+            // Next 2-4 bytes are the data
+            for(uint8_t j = 0; j < BMS_RESPONSE_LENGTH[i]; j++){
+                response[j+2] = BMS_RESPONSE_DATA[i][j];
+            }
+            // Send the message
+            uint8_t error = _can->sendMsgBuf(BMS_RESPONSE_ID, CAN_FRAME, BMS_RESPONSE_LENGTH[i]+2, response);
```

Comment:
I mentioned this in Teams but it may be worth adding a variable delay on the send here in order to emulate imperfect conditions. This was a problem during integration when we the telemetry system, so it'd be good to have a manual test for this in the firmware.

Reply:
I'll do some benchmarking with the BMS and flooding the bus with messages, and I'll see what kind of delay I should throw in.

Reply:
OK I did some testing with 10,000 messages per second (higher priority) being sent by the simulator while attempting to communicate directly with the BMS, and the delay went from 2ms up to about 3ms on average, with no more than about 5ms at the worst. I'll add a variable delay to simulate that.